syntax = "proto3";
option go_package = "./generated";

message TrainRequest {
  // path to train dataset
  string path = 1;
}

message DataSet {
  repeated Row rows = 1;
}

message Benchmark {
  int64 time = 1;
  int64 mem = 2;
}

message Row {
  repeated double x = 1;
  optional double y = 2;
}

message Model {
  repeated double weights = 1;
}

message TrainResponse {
  string modelId = 1;
  Benchmark benchmark = 2;
}

message PredictRequest {
  string modelId = 1;
  DataSet data = 2;
}

message PredictResponse {
  repeated double y = 1;
  Benchmark benchmark = 2;
}

service Ml {
  rpc train(TrainRequest) returns (TrainResponse);
  rpc predict(PredictRequest) returns (PredictResponse);
}

//protoc --go_out=./ --go-rpc_out=./ -I ./ ./*.proto